src/ur/workday.ur

```diff
@@ -0,0 +1,378 @@
+          debug ("Workday GET: " ^ prefix ^ url);
+          logged (WorldFfi.get (bless (prefix ^ url)) (WorldFfi.addHeader WorldFfi.emptyHeaders "Authorization" ("Bearer " ^ tok)) False)
+        end
+
+    fun apiOpt service url =
+        let val prefix = urlOfService service in
+          tok <- token;
+          debug ("Workday GET: " ^ prefix ^ url);
+          logged (WorldFfi.getOpt (bless (prefix ^ url)) (WorldFfi.addHeader WorldFfi.emptyHeaders "Authorization" ("Bearer " ^ tok)) False)
+        end
+
+    fun apiPost service url body =
+        let val prefix = urlOfService service in
+          tok <- token;
+          debug ("Workday POST: " ^ prefix ^ url);
+          debug ("Workday body: " ^ body);
+          logged (WorldFfi.post (bless (prefix ^ url)) (WorldFfi.addHeader WorldFfi.emptyHeaders "Authorization" ("Bearer " ^ tok)) (Some "application/json") body)
+        end
+
+    (* NOTE: could we improve performance by using tables instead of lists? do we need to? *)
+    (* NOTE: upstream this to a world library function *)
+    fun depaginate [object] (j : json object) (limit : int) (service : service) (endpoint : string) : transaction (list object) =
+        let fun depaginate' (offset : int) =
+          jsonResponse <- api service (endpoint ^ "?limit=" ^ (show limit) ^ "&offset=" ^ (show offset));
+          let val resp = (fromJson jsonResponse : response (list object))
+          in
+            if offset >= resp.Total
+            then return resp.Data
+            else
+                rest <- depaginate' (offset + limit);
+                return (List.append resp.Data rest)
+          end
+        in
+          depaginate' 0
+        end
+
+    (* Interface to Workday API for making GET and POST requests. *)
+    structure WorkdayApi = struct
+        structure WorkersApi = struct
+            fun get (workerId : wid) : transaction $worker =
+                s <- api Common ("/workers/" ^ workerId);
+                return (fromJson s : response $worker).Data
+                (* TODO: maybe this should return option worker? *)
+
+            val getAll : transaction (list $worker) =
+                depaginate 100 Common "/workers"
+        end
+
+        structure FeedbackApi = struct
+            fun get (workerId : wid) : transaction (list anytimeFeedback) =
+                depaginate 100 PerformanceEnablement ("/workers/" ^ workerId ^ "/anytimeFeedbackEvents")
+                (*return (
+                  (* Filter out feedback where the provider isn't in the workers table. *)
+                  (* HACK *)
+                  (* TODO: can we guarantee this doesn't happen? *)
+                  (*List.filter
+                    (fn fb => List.mem (Option.getOrError <xml>no</xml> fb.FromWorker).Id workers)*)
+                    (fromJson s : response (list anytimeFeedback)).Data)*)
+
+            fun getAll (workers : list wid) : transaction (list anytimeFeedback) =
+                List.mapConcatM get workers
```

Comment:
Oh dear, this is definitely a fundamentally slow implementation, with a whole line-up of sequential round trips to the server.

Reply:
Yup. We are currently looking into ways to avoid doing this as much as possible. It's a shame Workday does not expose an endpoint that returns feedback for many workers at once.